src/core/app.lua

```diff
@@ -283,18 +289,35 @@ function pace_breathing ()
    end
 end
 
+local jit_flush_if_needed = (function ()
+   local last_trigger = 0
+   return function (app)
+      local now = now()
+      if now < last_trigger then return end
+      if app.ingress_packet_drops then
+         local threshold = ingress_packet_drops.last_value + ingress_packet_drops.threshold
+         local current = app:ingress_packet_drops()
+         local exceeded = current > threshold
+         if exceeded then
+            ingress_packet_drops.last_value = current
+            last_trigger = now + ingress_packet_drops.wait
+            jit.flush()
+         end
+      end
+   end
+end)()
+
 function breathe ()
    monotonic_now = C.get_monotonic_time()
    -- Restart: restart dead apps
    restart_dead_apps()
    -- Inhale: pull work into the app network
    for i = 1, #app_array do
       local app = app_array[i]
---      if app.pull then
---         zone(app.zone) app:pull() zone()
       if app.pull and not app.dead then
          zone(app.zone)
          with_restart(app, app.pull)
+         jit_flush_if_needed(app)
```

Comment:
Diego I am so sorry to nitpick this thing to death, but I have another significant comment.

We're currently keeping a global counter for the "number of dropped packets the last time that we did a flush" value (`ingress_packet_drops.last_value`), but there can be more than one app, and their packet drop counts don't necessarily correlate.

Either we keep a per-app baseline for drops, or we sum drops across all apps.  I suggest we do the latter.  What do you think?  I suggest doing

```lua
local ingress_drop_monitor = {
   threshold = 100000,
   wait = 20,
   last_flush = 0,
   last_value = ffi.new('uint64_t[1]'),
   current_value = ffi.new('uint64_t[1]')
}

function ingress_drop_monitor:sample()
   local sum = self.current_value
   sum[0] = 0
   for i = 1, #app_array do
      local app = app_array[i]
      if app.pull and not app.dead then
         sum[0] = sum[0] + app:ingress_packet_drops()
      end
   end
end

function ingress_drop_monitor:jit_flush_if_needed()
   if self.current_value[0] - self.last_value[0] < self.threshold then return end
   if app.now() - self.last_flush < self.wait then return end
   self.last_flush = app.now()
   self.last_value[0] = self.current_value[0]
   jit.flush()
   --- in the future we change the last_flush, last_value, and current_value fields to be counters
end
``